source/base/hdf5.cc

```diff
@@ -0,0 +1,1752 @@
+          //  The C++ standards committee agreed to mandate that the storage
+          //  format used for the std::complex type be binary-compatible with
+          //  the C99 type, i.e. an array T[2] with consecutive real [0] and
+          //  imaginary [1] parts.
+          H5Tinsert(*t_type, "r", 0, H5T_NATIVE_DOUBLE);
+          H5Tinsert(*t_type, "i", sizeof(double), H5T_NATIVE_DOUBLE);
+          return t_type;
+        }
+
+      // The function should not reach this point
+      Assert(false, ExcInternalError());
+      return t_type;
+    }
+
+
+
+    // Several HDF5 functions such as H5Screate_simple() require a
+    // one-dimensional array that specifies the size of each dimension of the
+    // container, see:
+    // https://support.hdfgroup.org/HDF5/doc1.8/RM/RM_H5S.html#Dataspace-CreateSimple
+    // The function get_container_dimensions returns a vector with the
+    // dimensions of the container.
+    // For a std::vector the function returns std::vector<hsize_t>{vector_size}
+    // For a Vector the function returns std::vector<hsize_t>{vector_size}
+    // For a FullMatrix the function returns std::vector<hsize_t>{rows, columns}
+    //
+    // Instead of using functions with std::enable_if, "constexpr if" from C++17
+    // could be used
+    template <typename number>
+    std::vector<hsize_t>
+    get_container_dimensions(const std::vector<number> &data)
+    {
+      return std::vector<hsize_t>{data.size()};
+    }
+
+
+
+    template <typename number>
+    std::vector<hsize_t>
+    get_container_dimensions(const Vector<number> &data)
+    {
+      return std::vector<hsize_t>{data.size()};
+    }
+
+
+
+    template <typename number>
+    std::vector<hsize_t>
+    get_container_dimensions(const FullMatrix<number> &data)
+    {
+      return std::vector<hsize_t>{data.m(), data.n()};
+    }
+
+
+
+    // This function returns the total size of the container.
+    // For a std::vector the function returns int(vector_size)
+    // For a Vector the function returns int(vector_size)
+    // For a FullMatrix the function returns int(rows*columns)
+    template <typename number>
+    unsigned int
+    get_container_size(const std::vector<number> &data)
+    {
+      return static_cast<unsigned int>(data.size());
+    }
+
+
+
+    template <typename number>
+    unsigned int
+    get_container_size(const Vector<number> &data)
+    {
+      return static_cast<unsigned int>(data.size());
+    }
+
+
+
+    template <typename number>
+    unsigned int
+    get_container_size(const FullMatrix<number> &data)
+    {
+      return static_cast<unsigned int>(data.m() * data.n());
+    }
+
+
+
+    // This function returns the pointer to the raw data of a container
+    template <typename number>
+    void *
+    get_container_pointer(std::vector<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong
+      return data.data();
```

Comment:
that is true, but I think obvious. I would remove the comment, but if you want to keep it, put it into the comment of the function itself, not of the statement below where I think it is just confusing.